src/ssd/ssd-titlebar.c

```diff
@@ -241,6 +244,58 @@ set_alt_button_icon(struct ssd *ssd, enum ssd_part_type type, bool enable)
 	} FOR_EACH_END
 }
 
+/*
+ * Usually this function sets ssd->state.visible_buttons_count_{left,right} to
+ * wl_list_length(&rc.title_buttons_{left,right}), but smaller numbers can be
+ * set for tiny windows.
+ */
+static void
+update_visible_buttons(struct ssd *ssd)
+{
+	struct view *view = ssd->view;
+	int width = view->current.width;
+	int button_width = view->server->theme->window_button_width;
+	int button_count_left = wl_list_length(&rc.title_buttons_left);
+	int button_count_right = wl_list_length(&rc.title_buttons_right);
+
+	/*
+	 * The corner-left button is lastly removed as it's usually a window
+	 * menu or app icon button.
+	 */
+	while (width < button_width * (button_count_left + button_count_right)) {
+		if (button_count_left > button_count_right) {
+			button_count_left--;
+		} else {
+			button_count_right--;
+		}
+	}
+
+	int button_count;
+	struct ssd_part *part;
+	struct ssd_sub_tree *subtree;
+	struct title_button *b;
+	FOR_EACH_STATE(ssd, subtree) {
+		button_count = 0;
+		wl_list_for_each(b, &rc.title_buttons_left, link) {
+			part = ssd_get_part(&subtree->parts, b->type);
+			wlr_scene_node_set_enabled(part->node,
+				button_count < button_count_left);
+			button_count++;
+		}
+
+		button_count = 0;
+		wl_list_for_each_reverse(b, &rc.title_buttons_right, link) {
+			part = ssd_get_part(&subtree->parts, b->type);
+			wlr_scene_node_set_enabled(part->node,
+				button_count < button_count_right);
+			button_count++;
+		}
+	} FOR_EACH_END
+
+	ssd->state.visible_button_count_left = button_count_left;
+	ssd->state.visible_button_count_right = button_count_right;
```

Comment:
I understand the remaining de-sync issue, but still don't like caching these. Would looping through the SSD buttons and counting whether they are enabled in the scene be terrible?

Reply:
Not terrible at all. It's just my weak preference and I understand adding states to `struct ssd` for this minor feature is undesirable.

I made a fixup commit that counts the number of enabled button nodes instead of caching it.
I still slightly prefer caching it as it's less verbose, but I'll squash it if you like it.

Reply:
I think the uncached version is better, but we could see what the others think.

Reply:
Yeah, I much prefer the un-cached version (fixup-commit).
Caching is always harder to maintain and given the context, a few extra loops is highly unlikely to be expensive.
Thanks for having the discussion on this.

Reply:
Alright. I squashed the commit with some minor improvements. Thank your for your advises!